Add Intl.DateTimeFormat.resolvedOptions dateStyle, timeStyle

medmunds opened the pull request:
<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

Document `dateStyle` and `timeStyle` for [`Intl.DateTimeFormat.resolvedOptions()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/resolvedOptions#description) return value.

### Motivation

These properties were added in https://github.com/tc39/ecma402/pull/457, but are not covered by the current documentation.

### Additional details

See current https://tc39.es/ecma402/#table-datetimeformat-resolvedoptions-properties


### Related issues and pull requests

<!-- 🔨 If this fully resolves a GitHub issue, use "Fixes #123" -->
<!-- 👉 Highlight related pull requests using "Relates to #123" -->
<!-- ❗ If another pull request should be merged first, use "**Depends on:** #123" -->


<!-- 👷‍♀️ After submitting, go to the "Checks" tab of your PR for the build status -->

medmunds committed:
Document `dateStyle`, `timeStyle` resolved options

(See current https://tc39.es/ecma402/#table-datetimeformat-resolvedoptions-properties)

medmunds committed:
Lint

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Josh-Cena commented on files/en-us/web/javascript/reference/global_objects/intl/datetimeformat/resolvedoptions/index.md:
```diff
@@ -52,6 +52,10 @@ The resulting object has the following properties:
     not be represented in formatted output.
+- `dateStyle`, `timeStyle`
+  - : The values provided for these properties in the `options` argument, if any.
+    (The `dateStyle` and `timeStyle` shortcuts are _not_ resolved to individual
+    date-time format properties like `year` or `hour`.)
```
I'm not sure I understand this sentence. Do you mean that if I provide `{ dateStyle: "full" }`, I would not expect `year: "full", month: "full", day: "full"`, but rather them being absent? We never said `dateStyle: "full"` is a shortcut for `year: "full", month: "full", day: "full"` anyway, and I doubt that's always the case in practice.

medmunds replied:
I struggled with wording for that, and am hoping you might come up with something clear but succinct. Here's the verbose explanation:

`{dateStyle: "medium"}` is a shortcut for some [locale-and-calendar-specific combination](https://cldr.unicode.org/translation/date-time/date-time-patterns#h.aa5zjyepm6vh) of era, year, month, day, and weekday options. (In en-US, the "medium" dateStyle is equivalent to using `{year: "numeric", month: "short", day: "numeric"}`; for ja-JP you'd need to substitute `month: "numeric"` to get the matching "medium" format.)

As a developer, I might reasonably expect that the `resolvedOptions()` method would "resolve" (expand) the `dateStyle` and `timeStyle` shortcuts to the specific date and time options for the locale, similar to how it resolves the locale's preferred calendar. But I would be wrong:

```javascript
// Let's find the preferred calendar and date format for Thailand:
let format = new Intl.DateTimeFormat("th-TH", {dateStyle: "medium"});
let {calendar, year, month, day} = format.resolvedOptions();
console.log(calendar); 
// 'buddhist'  (that's helpful!)
console.log(year, month, day); 
// undefined, undefined, undefined  (oops, that didn't work)
```

[Incidentally, it seems like this behavior was briefly considered, but rejected, as part of the dateStyle/timeStyle proposal: tc39/proposal-intl-datetime-style#49.]

Josh-Cena replied:
I see. That's really helpful. I would suggest:

1. We modify the `Intl.DateTimeFormat` constructor doc too, to mention that `dateStyle` and `timeStyle` are shortcuts for which properties, and that the exact datetime component style values they expand to may be different depending on the locale;
2. Here, we say: "Although `dateStyle` and `timeStyle` are shortcuts for individual datetime component styles, the exact component styles they resolve to are not included in the resolved options. This is because an options object with both `dateStyle` and individual date component styles is not a valid options object for the `Intl.DateTimeFormat` constructor."
3. We add your example with the expanded calendar to this page. You can add other examples to the constructor page too, if you find them helpful—I don't think we do a great job at documenting `dateStyle` and `timeStyle` because I'm not as familiar with Intl.

medmunds replied:
I think that all sounds great.

For 1, perhaps wording like "expands internally" or something that suggest the details are not exposed.

Incidentally, the "Try It" example on this page does actually show an example of resolving the calendar for a locale. (The other common use of resolvedOptions is to get the local (system) time zone: `new Intl.DateTimeFormat().resolvedOptions().timeZone`, and that might be a useful example to add.)

Josh-Cena replied:
You can go ahead with wordings you prefer—I don't have strong opinions.

medmunds replied:
Done. (For this page. I'll let you handle updates to the constructor page separately.)

Josh-Cena reviewed (commented):
Thank you! I'm in the process of updating these docs but it's good to make them more correct. I just have one question.

medmunds committed:
Update from review feedback

- Clarify dateStyle and timeStyle
- Add example
- Add missing hourCycle
- Clarify calendar and numberingSystem

medmunds committed:
lint

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Josh-Cena committed:
Update index.md

Josh-Cena committed:
Add example

Josh-Cena reviewed (approved):
Hi, thanks! As I mentioned, I already have stashed work locally that aims to unify the format of all `resolvedOptions` pages, so I took this opportunity to apply that format. I have made sure to incorporate your changes.

Also as suggested I updated the constructor page slightly. I hope everything looks good!

medmunds commented on files/en-us/web/javascript/reference/global_objects/intl/datetimeformat/datetimeformat/index.md:
```diff
@@ -148,9 +148,9 @@ The default value for each date-time component option is {{jsxref("undefined")}}
 
 - `dateStyle`
-  - : The date formatting style to use when calling `format()`. Possible values are `"full"`, `"long"`, `"medium"`, and `"short"`.
+  - : The [date formatting style](https://cldr.unicode.org/translation/date-time/date-time-patterns#h.aa5zjyepm6vh) to use. Possible values are `"full"`, `"long"`, `"medium"`, and `"short"`. It resolves to styles for `weekday`, `day`, `month`, `year`, and `era`, with the exact combination of values depending on the locale.
```
"resolves" is potentially misleading here, since there's a `resolvedOptions` function that *won't* actually tell you what the shortcut resolved to. (Maybe "expands" or "results in" or ???)

medmunds commented on files/en-us/web/javascript/reference/global_objects/intl/datetimeformat/datetimeformat/index.md:
```diff
@@ -215,13 +217,19 @@ console.log(shortTime.format(Date.now())); // "1:31 PM"
   dateStyle: "short",
 });
-console.log(shortDate.format(Date.now())); // "07/07/20"
+console.log(shortDate.format(Date.now())); // "7/7/20"
```
The exact output here depends on how `"en"` resolves in your browser. If it resolves to (e.g.,) `"en-UK"`, you'll get "07/07/20" as in the original example. If it resolves to "en-US", you'll get "7/7/20" as in your edit. (Also, the day and month would swap places.)

Maybe use `"en-US"` in the examples? Or add a comment that this example is running on an en-US system? Or show two examples for each result?

(Similarly, the shortTime will result in either "1:31 PM" or "13:31" depending on what locale `"en"` resolves to. The mediumTime example below will also vary.)

Josh-Cena replied:
I was under the impression `"en"` always resolves to `"en-US"` because that's how Add Likely Subtags works, but I'm very curious if that's not the case. In the meantime I've added `-US` for least ambiguity.

medmunds replied:
Oh, sorry, you're absolutely right—it uses the "implementation's" default locale, not the (user's) system locale setting (or browser's language setting).

(Which means I need to add `navigator.languages` to my resolvedOptions example.)

Josh-Cena replied:
@medmunds It does default to the user's system locale! See https://tc39.es/ecma402/#sec-defaultlocale: 

> The returned String value represents the structurally valid ([6.2.1](https://tc39.es/ecma402/#sec-isstructurallyvalidlanguagetag)) and canonicalized ([6.2.2](https://tc39.es/ecma402/#sec-canonicalizeunicodelocaleid)) [language tag](https://tc39.es/ecma402/#sec-language-tags) for the [host environment](https://tc39.es/ecma262/#host-environment)'s current locale.

medmunds replied:
And I'm in en-US, so have that bias. (I frequently adjust my browser language preferences for testing, but not my system locale.)

It looks like behavior of `new Intl.DateTimeFormat("en")` comes down to the implementation of [LookupMatchingLocaleByBestFit](https://tc39.es/ecma402/#sec-lookupmatchinglocalebybestfit), which is "implementation dependent, but should produce results that a typical user of the requested locales would consider at least as good as those produced by the LookupMatchingLocaleByPrefix algorithm." And it looks like ByPrefix isn't permitted to expand "en" to "en-UK" (or more likely "[en-001](https://cldr.unicode.org/development/development-process/design-proposals/english-inheritance)"), but I can't tell if that's true of ByBestFit.

medmunds reviewed (commented):
Much better wording, thanks! Flagged a couple of potential issues inline.

Josh-Cena committed:
Reviews

medmunds committed:
Fix locale preferences example

medmunds committed:
Revert 5b639ad56b72014a2359c55d642ea4821a42bf9f

Josh-Cena committed:
Remove wrapping

Josh-Cena commented:
Is this ready for merging?

medmunds commented:
> Is this ready for merging?

I think so. Thanks!

Josh-Cena commented:
Great, welcome aboard! 😄
